sig.go

```diff
@@ -0,0 +1,96 @@
+package rmb
+
+import (
+	"crypto/ed25519"
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+
+	sr25519 "github.com/ChainSafe/go-schnorrkel"
+	"github.com/threefoldtech/substrate-client"
+
+	"github.com/gtank/merlin"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	SignatureTypeEd25519 = "ed25519"
+	SignatureTypeSr25519 = "sr25519"
+)
+
+type Verifier interface {
+	Verify(msg []byte, sig []byte) bool
+}
+
+type Ed25519VerifyingKey []byte
+type Sr25519VerifyingKey []byte
+
+func (k Ed25519VerifyingKey) Verify(msg []byte, sig []byte) bool {
+	return ed25519.Verify([]byte(k), msg, sig)
+}
+
+func signingContext(msg []byte) *merlin.Transcript {
+	return sr25519.NewSigningContext([]byte("substrate"), msg)
+}
+
+func (k Sr25519VerifyingKey) verify(pub sr25519.PublicKey, msg []byte, signature []byte) bool {
+	var sigs [64]byte
+	copy(sigs[:], signature)
+	sig := new(sr25519.Signature)
+	if err := sig.Decode(sigs); err != nil {
+		return false
+	}
+	return pub.Verify(sig, signingContext(msg))
+}
+
+func (k Sr25519VerifyingKey) pubKey() (*sr25519.PublicKey, error) {
+	var pubBytes [32]byte
+	copy(pubBytes[:], k)
+	pk := new(sr25519.PublicKey)
+
+	if err := pk.Decode(pubBytes); err != nil {
+		return nil, err
+	}
+	return pk, nil
+}
+
+func (k Sr25519VerifyingKey) Verify(msg []byte, sig []byte) bool {
+	pk, err := k.pubKey()
+	if err != nil {
+		log.Error().Str("pk", hex.EncodeToString(k)).Err(err).Msg("failed to get sr25519 key from bytes returned from substrate")
+		return false
+	}
+	return k.verify(*pk, msg, sig)
+}
+
+func challenge(m *Message) ([]byte, error) {
```

Comment:
I think the challenge must also take other parts of the message into account, this include Version, ID, TwinSrc, and TwinDst are also important, otherwise a hacker can rerun the same payload again.
I even think we can include the Epoch into the hash, then the server can reject messages older than a certain age (let's say 20 seconds) to prevent reply attacks.

Of course if the epoch is taken into account, the challenge must be recalculated on sent (on each retry) to use the correct time.